src/util/strconv.h

```diff
@@ -115,57 +116,64 @@ static char *rm_strndup_unescape(const char *s, size_t len) {
 // returns the bytes written to encoded, or 0 if the length of the transformed
 // string is greater than in_len and no transformation was done
 static size_t unicode_tolower(char *encoded, size_t in_len) {
-  uint32_t u_stack_buffer[SSO_MAX_LENGTH];
-  uint32_t *u_buffer = u_stack_buffer;
-
   if (in_len == 0) {
     return 0;
   }
 
-  const char *encoded_char = encoded;
+  uint32_t u_stack_buffer[SSO_MAX_LENGTH];
+  uint32_t *u_buffer = u_stack_buffer;
+
   ssize_t u_len = nu_strtransformnlen(encoded, in_len, nu_utf8_read,
                                               nu_tolower, nu_casemap_read);
 
-  if (u_len >= (SSO_MAX_LENGTH - 1)) {
-    u_buffer = (uint32_t *)rm_malloc(sizeof(*u_buffer) * (u_len + 1));
+  if (u_len > (SSO_MAX_LENGTH - 1)) {
+    u_buffer = (uint32_t *)rm_malloc(sizeof(uint32_t) * (u_len + 1));
   }
 
   // Decode utf8 string into Unicode codepoints and transform to lower
-  uint32_t codepoint;
+  const char *encoded_char = encoded;
   unsigned i = 0;
-  for (ssize_t j = 0; j < u_len; j++) {
+  while (encoded_char < encoded + in_len) {
+    uint32_t codepoint = 0;
     // Read unicode codepoint from utf8 string
+    // This might read more than one char.
     encoded_char = nu_utf8_read(encoded_char, &codepoint);
+    if (codepoint == 0) {
+      // If we reach the end of the string, break
+      break;
+    }
     // Transform unicode codepoint to lower case
     const char *map = nu_tolower(codepoint);
 
     // Read the transformed codepoint and store it in the unicode buffer
-    // map would be NULL if no transformation is needed,
-    // i.e.: lower case is the same as the original, emoji, etc.
     if (map != NULL) {
       uint32_t mu;
       while (1) {
         map = nu_casemap_read(map, &mu);
         if (mu == 0) {
           break;
         }
-        u_buffer[i] = mu;
-        ++i;
+        u_buffer[i++] = mu;
       }
-    }
-    else {
+    } else {
       // If no transformation is needed, just copy the unicode codepoint
-      u_buffer[i] = codepoint;
-      ++i;
+      u_buffer[i++] = codepoint;
     }
   }
-
+  RS_LOG_ASSERT_FMT(i <= u_len, "i (%u) should be less equal to u_len (%zd)", i, u_len);
   // Encode Unicode codepoints back to utf8 string
   ssize_t reencoded_len = nu_bytenlen(u_buffer, u_len, nu_utf8_write);
   if (reencoded_len > 0 && reencoded_len <= in_len) {
     nu_writenstr(u_buffer, u_len, encoded, nu_utf8_write);
   } else {
     reencoded_len = 0;
+    // reallocate the string to fit the transformed string
+    // rm_free(encoded);
+    // encoded = (char *)rm_malloc(reencoded_len + 1);
+    // nu_writenstr(u_buffer, u_len, encoded, nu_utf8_write);
+
+    // encoded[reencoded_len] = '\0'; // Null-terminate the string
+    // // Update the length to the new length
```

Comment:
Remove and consider adding a comment to explain why we would enter the else (or elaborate the documentation at the header regarding return value = 0).